Migrate Nextflow examples to public repository.

emmarogge opened the pull request.

emmarogge committed:
Move Nextflow notebook to public repository.

emmarogge committed:
Add README.

emmarogge committed:
Refactor content for clarity.

emmarogge committed:
Add links to notebook snapshots so users can preview expected results.

emmarogge committed:
Provide notebook preview links in the README.

emmarogge committed:
Fix link in README.

amygdala commented on nextflow/nextflow_examples.ipynb:
```diff
@@ -0,0 +1,1108 @@
+   "id": "543efb9d-6e18-4677-a002-4429f58fd947",
+   "metadata": {},
+   "source": [
+    "On Terra cloud environments, git is configured to transform any `https` URL into  a `git@github.com` URL; the user's SSH key cannot be derived from the `https` URL, and therefore access to any private repo requires the `git@github.com` URL instead. However, this global setting prevents users who haven't set up their Terra SSH key from accessing public GitHub repositories. Therefore, **if you ...
```
I'm guessing that https://broadworkbench.atlassian.net/browse/PF-2628 should be in prod soon (like next week) and we can take this out.  Fine to wait until it's actually in place.

amygdala commented on nextflow/nextflow_examples.ipynb:
```diff
@@ -0,0 +1,1108 @@
+    "</div>\n",
+    "\n",
+    "**If your workspace is a clone of the Getting Started with Workflows workspace...**\n",
+    "    Two buckets with the workspace reference names 'ws_files' and 'ws_files_autodelete_after_two_weeks' have been created in your workspace, and the contents of each respective bucket in the parent workspace has been copied over to the corresponding bucket in this workspace. For more details on which resources are created or copied when a workspace is cloned, please see the [Terra CLI docume ...
```
With the changes in the workspace_setup notebook, these buckets no longer get cloned, so they'll need to run this notebook either way now.

emmarogge replied:
Fixed!

amygdala commented on nextflow/nextflow_examples.ipynb:
add a bit of info here on how they find the session ID. (I'm guessing it just comes from `nextflow log` but you will need to spell it out for them). Potentially you could move the 'nextflow log' cell to above this one if that makes sense for your flow.

emmarogge replied:
Fixed -- I put the logging section before this section and added a hint.

amygdala commented on nextflow/nextflow_examples.ipynb:
```diff
@@ -0,0 +1,1108 @@
+    "tags": []
+   },
+   "source": [
+    "# Nextflow in Terra"
```
Add the top matter links as in https://github.com/verily-src/terra-solutions-mc-terra-testing/blob/main/notebook_template.ipynb
<img width="609" alt="CleanShot 2023-04-24 at 16 49 51@2x" src="https://user-images.githubusercontent.com/115093/234139387-f86f0803-7f85-432a-a2cf-9e7afa7403cb.png">

emmarogge replied:
done!

amygdala reviewed (commented):
I will give the nb a test run tomorrow, but thought I would add a few comments in case you want to tackle them tomorrow morning.

amygdala commented on nextflow/nextflow_examples.ipynb:
I'm still seeing the following error.  Are you able to dig into this a bit more? I'm guessing you may want to remove some fields from the config ... or something along those lines. In a pinch you could set that param to ignore unknown configs, but it would be better to fix so that the issue doesn't occur. (It's presumably harmless but suggests we don't quite know what we're doing :)
<img width="1207" alt="CleanShot 2023-04-25 at 09 39 25@2x" src="https://user-images.githubusercontent.com/115093/234346176-1c670829-a2ae-41c8-9c96-9eaebfb3959d.png">

emmarogge replied:
It's actually just a warning, not an error, and it cannot be disabled. Since there's a canonical config file for all `google` in `nf-core`, this is simply alerting the user that they're overriding the defaults here. (Which we definitely want them to do, for example -- we don't want the default zone, `europe-west2-c`!)

amygdala replied:
I figured out a way to address with a different version of the config file that AFAICT is working properly-- let's talk more on Monday and I'll show you what I did.

emmarogge replied:
fixed!

amygdala commented on nextflow/nextflow_examples.ipynb:
With the 'resume' (which I've never actually tested until now :), I see the following error.  Does this mean that the outdir should be changed to a new path for the 'resume' run?
![CleanShot 2023-04-25 at 11 19 55@2x](https://user-images.githubusercontent.com/115093/234367294-9d264296-3e0b-4c0e-b26d-cecab21b3b5a.png)

emmarogge replied:
I've added content explaining this behavior, which is a known Nextflow bug, and validated that changing the output directory resolves this issue, as well as providing a different output directory in the command. See [screenshot](https://screenshot.googleplex.com/8pUW9omemGRFgzh) for validation that using a different directory results in successful execution/no warning.

amygdala replied:
If you just change the outdir in the 'resume' example (e.g. change it to `.../nf_core/output_resume` or something, what happens?

emmarogge committed:
Number the examples & relocate cloning of rnaseq-nf repo to Example 1 section.

emmarogge committed:
Provide cost estimates for examples.

emmarogge committed:
Add workspace description.

emmarogge committed:
Fix section about workspace_setup.ipynb.

emmarogge committed:
Reorder nf-core section to put logging before resuming workflow execution.

emmarogge committed:
Add GitHub & Vertex AI links to header.

emmarogge committed:
Describe known Nextflow bug causing error if same output dir is used when resuming execution.

amygdala commented on nextflow/workspace_description.md:
```diff
@@ -0,0 +1,54 @@
+
+Two examples are provided in `nextflow_examples.ipynb`. You will find this notebook in [https://github.com/DataBiosphere/terra-axon-examples/tree/main/nextflow](https://github.com/https://github.com/DataBiosphere/terra-axon-examples/tree/main/nextflow).  As mentioned above, its repo, `terra-axon-examples`, will be automatically cloned to your Enterprise Terra Cloud Environment, and you should be  ...
+- The first example demonstrates running a Nextflow RNASeq workflow on human gut data.
+- The second example demonstrates running the `nf-core` RNASeq pipeline on yeast genome data.
```
link to nf-core where you first mention it (https://nf-co.re/)

emmarogge replied:
done!

amygdala commented on nextflow/workspace_description.md:
```diff
@@ -0,0 +1,54 @@
+  Then, [add that key](https://docs.github.com/en/authentication/connecting-to-github-with-ssh/adding-a-new-ssh-key-to-your-github-account) to your GitHub account.
+  [This page](https://terra-docs.api.verily.com/docs/how_to_guides/terra_ssh_key_guide/) has more detail.
+
+  This setup allows the `terra-axon-examples` example repo, which you can see listed as a Git repository for the workflow, to be automatically cloned when you create a Cloud Environment.
```
This is too specific— rephrase to indicate that it will allow all the (three) repos to be cloned.
(side note: hopefully the rollback will make into prod next week; then we can change this)

emmarogge replied:
fixed!

amygdala commented on nextflow/workspace_description.md:
```diff
@@ -0,0 +1,54 @@
+
+Previews of this notebook, including outputs, are provided in the workspace resource folder "Notebook snapshots". Click the links listed below to view notebook previews in order to gain a better understanding of the setup and analysis included in this demonstration workspace:
+
+1. [Workspace setup] (https://terra-preprod-ui-terra.api.verily.com/workspaces/getting-started-with-workflows-workspace/resources/97ce452a-dd3c-4045-97d3-bc70db429f83/notebook_snapshots/workspace_setup.html)
```
It looks like the rogue space is preventing this from rendering properly.

emmarogge replied:
fixed!

amygdala commented on nextflow/workspace_description.md:
```diff
@@ -0,0 +1,54 @@
+Previews of this notebook, including outputs, are provided in the workspace resource folder "Notebook snapshots". Click the links listed below to view notebook previews in order to gain a better understanding of the setup and analysis included in this demonstration workspace:
+
+1. [Workspace setup] (https://terra-preprod-ui-terra.api.verily.com/workspaces/getting-started-with-workflows-workspace/resources/97ce452a-dd3c-4045-97d3-bc70db429f83/notebook_snapshots/workspace_setup.html)
+2. [Example 1: Run a Nextflow workflow](https://terra-preprod-ui-terra.api.verily.com/workspaces/getting-started-with-workflows-workspace/resources/21ffa484-75b9-46ff-b70d-6ebeb2962eb2/notebook_snapshots/nextflow_examples_first_exercise_fully_executed.html)
```
It's very cool that these links work!  However, when you click, something happens that will be confusing to the user.  Their browser window will switch to the original workspace, not the one they cloned. Since these workspaces will look nearly identical, the avg user won't notice, and this will lead to issues.  We can't use inline html in these descriptions and so I don't believe there's a way to open these links in a new window.

Let's talk about this more. Maybe some kind of relative link is possible.  Or disappointingly it might work better overall to just direct them to navigate to the "Resources" panel, & maybe include a screenshot that indicates the folder to click on.

emmarogge replied:
I think including the preview links is beneficial to users overall even if the navigation back to the user's clone workspace is tedious. If users were to click the links from a clone of the parent workspace, little harm could follow as the parent workspace will be read-only. Even if users become confused and attempt to change/interact with the parent workspace, they will not be able, and will realize they need to navigate back to their clone of the parent workspace. The only other downside I can imagine might be that users end up with multiple clones of the same workspace, but they'd have to be pretty clueless for this to occur and such a mistake is easily fixed by the user deleting the unnecessary workspaces. I do not think this, on balance, is a problem, but if there are other downstream issues this approach could cause, I am open to refactoring. Perhaps this is a question we should take to the broader team?

I do agree that relative linking and/or support for  HTML in the workspace description would be helpful.

wnojopra replied:
I agree with Amy that we shouldn't link directly to the resource in the original workspace. Since relative links aren't possible at the moment, I also agree that directing the user to navigate to the Resources panel would be better.

emmarogge replied:
noted -- I have refactored this section accordingly

amygdala commented on nextflow/workspace_description.md:
````diff
@@ -0,0 +1,54 @@
+> **If you have not set up your ssh key, and do not see the `terra-axon-examples` subdirectory**, first open a Terminal window on the notebook server (under the **File** menu) and run:
+
+  ```sh
+  git clone https://github.com/DataBiosphere/terra-axon-examples.git
````
side note: currently, if they don't set up the key and see the repo, they'll also need to run `git config --unset --global url."git@github.com:".insteadOf` .  You have this in the notebook, but it might be worth mentioning here too. (I need to add this to the other workspace overview as well).    
But, that should soon be unnecessary, and at that time we'll want to rephrase this whole discussion, in that they don't _need_ the ssh key set up for the repos used in this example.

emmarogge replied:
Now that the change has landed in prod, I've refactored this to remove the SSH key content.

amygdala commented on nextflow/workspace_description.md:
```diff
@@ -0,0 +1,54 @@
+- The first example demonstrates running a Nextflow RNASeq workflow on human gut data.
+- The second example demonstrates running the `nf-core` RNASeq pipeline on yeast genome data.
+
+The examples provided are independent of one another (meaning you can run only the second example if desired, without having to execute the first example).
```
You should probably add that they need to run the setup section (whatever is the right way to refer to it) regardless, else some users will inevitably just skip down to the second example w/out doing the setup.

emmarogge replied:
Good point; I've added that the setup section must be run before either example.

amygdala commented on nextflow/nextflow_examples.ipynb:
```diff
@@ -0,0 +1,1132 @@
+    "# Nextflow in Terra\n",
+    "<table align=\"left\">\n",
+    "<td>\n",
+    "<a href=\"https://github.com/DataBiosphere/terra-axon-examples/blob/dea53d1d347b7c81311c766f6bd66b06a3025752/nextflow/nextflow_examples.ipynb\">\n",
```
thanks for adding these!  We'll want to change them to point to the 'main' branch.  It probably works best to go ahead and do that now, it's fine if they don't work until merged in.

emmarogge replied:
Done!

amygdala commented on nextflow/nextflow_examples.ipynb:
IIUC, the nf-core pipeline maintainers and community are independent from the team actually building Nextflow. I know it seems like a small thing but we should be careful of attribution.
Also, add a link (https://nf-co.re/)

emmarogge replied:
Yes, that's why I link to it in the latter sentence of this paragraph.

amygdala commented on nextflow/nextflow_examples.ipynb:
This arg (minAssignedFrags) causes a warning, and I believe it's b/c it's actually a `salmon` arg.  I don't think it's doing anything in this call and can be removed .

emmarogge replied:
This is outdated -- this parameter isn't used any longer.

emmarogge committed:
Remove workaround for git@ URLs now that fix has reached prod.

emmarogge committed:
Instruct users to run setup section before examples

emmarogge committed:
Point to main branch of repo for header links.

emmarogge committed:
Remove git@ URL config since it is no longer needed.

emmarogge committed:
Refactor nf-core RNASeq config to prevent 'unused parameters' warning.

emmarogge committed:
Remove links to notebook previews due to Amy's concern that it may confuse users to view the master/parent workspace Resources tab.

amygdala commented on nextflow/workspace_description.md:
```diff
@@ -0,0 +1,46 @@
+- *Duplicate* this workspace. You can do that via the 'three-dot' menu in the upper right of the workspace. Duplicating this workspace ensures three Git repositories required for these examples will be automatically cloned when you create a Cloud Environment.
+
+   The required repositories which will be cloned include:
+  1. (`terra-axon-examples` example repo)[https://github.com/DataBiosphere/terra-axon-examples]: Contains JupyterLab notebook `nextflow_examples.ipynb` which demonstrates how to configure and run Nextflow workflows in Enterprise Terra.
```
somehow the link syntax fixes didn't make it into the checked-in version.

emmarogge replied:
this has been fixed

emmarogge replied:
this has since been fixed: https://screenshot.googleplex.com/7dTYko2o6EZoB4q

amygdala commented on nextflow/workspace_description.md:
```diff
@@ -0,0 +1,46 @@
+
+Launch the environment once it's running, and then run the notebook [workspace_setup.ipynb](https://github.com/DataBiosphere/terra-axon-examples/blob/main/workspace_setup.ipynb). Its repo, `terra-axon-examples`, which is defined as a workspace Git repository, should be automatically cloned to your Enterprise Terra Cloud Environments, and you should be able to navigate to the notebook in the Jupyt ...
+
+> **If you have not set up your ssh key, and do not see the `terra-axon-examples` subdirectory**, first open a Terminal window on the notebook server (under the **File** menu) and run:
```
I'd remove the "if you have not set up your ssh key" part -- with the rollback, and your change to define the workspace git refs using https, the ssh key should not matter.  (IMO It's still useful to include the 'git clone' instruction though, just in case).

emmarogge replied:
I've removed the SSH key clause and left this: https://screenshot.googleplex.com/9nxrPRvx65BjaEM

amygdala commented on nextflow/workspace_description.md:
```diff
@@ -0,0 +1,46 @@
+
+Create an Enterprise Terra Cloud environment, by navigating to the "Environments" tab of the workspace. You can use the configuration defaults.
+
+Launch the environment once it's running, and then run the notebook [workspace_setup.ipynb](https://github.com/DataBiosphere/terra-axon-examples/blob/main/workspace_setup.ipynb). Its repo, `terra-axon-examples`, which is defined as a workspace Git repository, should be automatically cloned to your Enterprise Terra Cloud Environments, and you should be able to navigate to the notebook in the Jupyt ...
```
Now that I've (annoyingly :) just merged the setup notebook refactoring, we should rephrase this.  I was going to offer to just do it myself as a following PR, but since we need a few last changes to this description file, we can go ahead and add it to this PR.  I'll talk to you more about this tomorrow.

amygdala commented on nextflow/workspace_description.md:
```diff
@@ -0,0 +1,46 @@
+Two examples are provided in `nextflow_examples.ipynb`. You will find this notebook in [https://github.com/DataBiosphere/terra-axon-examples/tree/main/nextflow](https://github.com/https://github.com/DataBiosphere/terra-axon-examples/tree/main/nextflow).  As mentioned above, its repo, `terra-axon-examples`, will be automatically cloned to your Enterprise Terra Cloud Environment, and you should be  ...
+
+- The first example demonstrates running a Nextflow RNASeq workflow on human gut data.
+- The second example demonstrates running the (`nf-core`)[https://nf-co.re/]RNASeq pipeline on yeast genome data.
```
this link needs the syntax fixed as well.

emmarogge replied:
done

amygdala commented on nextflow/nextflow_examples.ipynb:
```diff
@@ -0,0 +1,1169 @@
+   "source": [
+    "### Notebook setup <a id=\"workspace_setup\"> \n",
+    "<div class=\"alert alert-block alert-info\">\n",
+    "<b>Note:</b> This notebook assumes that you have run <a href=\"../workspace_setup.ipynb\">workspace_setup.ipynb</a> in the parent directory to create necessary workspace resources. Please note that you can skip creating a BigQuery dataset as that won't be used in this notebook's examples.\n",
```
same comment as in the description.md that now that the setup nb refactoring is in, we need to change this text, sorry about that.  I can do it as a following PR if you'd prefer.

amygdala commented on nextflow/nextflow_examples.ipynb:
It looks like you still need to remove the older version of this from the notebook.

emmarogge replied:
done

amygdala commented on nextflow/nextflow_examples.ipynb:
```diff
@@ -0,0 +1,1169 @@
+   },
+   "outputs": [],
+   "source": [
+    "!terra resource resolve --name=rnaseq-nf-repo || (terra resource add-ref git-repo \\\n",
```
See [this](https://chat.google.com/room/AAAA7Xi94tc/8iPv9SnXspY) chat discussion -- we're going to have to change things one direction or the other (either change the resource name to be the same as the directory you expect to exist, or change the cells to expect a dir named for the resource name).  We'll need to do the same for 'test-datasets'.
We can talk about this more tomorrow.

emmarogge replied:
I chose to update the names of the resources to the actual names of the repositories, thus can keep the directory names the same.

amygdala commented on nextflow/nextflow_examples.ipynb:
Test this from a fresh notebook VM.  This is no longer working for me unless I restart the kernel. Apparently something has changed.
We probably should:
1. move this import to the top of the setup section (I was thinking to suggest that anyway)
2. split up the actions: do the install, then add a cell to restart the kernel (that's why we want this at the start of the setup)
3. then do the import

wnojopra replied:
I'm also seeing errors here. In my experience, package installs require a kernel restart before it can be imported. I'd recommend implementing one of Matt's or Amy's suggestions from the chat here. In particular, I'm fond of moving this import to the top, and then explicitly say "after running it, you'll likely need to restart your kernel".

wnojopra replied:
Thanks for updating this, Emma. Could you please move the imports all the way to the top? Right now, the BUCKET_REFERENCE cell is above the imports. So after I restart the kernel, I lose the BUCKET_REFERENCE variable.

amygdala commented on nextflow/nextflow_examples.ipynb:
This needs the `gls` profile instead here as well. (same as with the initial run)

emmarogge replied:
Fixed.

amygdala commented on nextflow/nextflow_examples.ipynb:
I believe that this arg (which as I mentioned earlier I think is related to salmon) doesn't do anything here but cause a warning and should be removed.  (same as with the initial run)

emmarogge replied:
This has been removed.

amygdala reviewed (commented):
I found a few things that still need to be cleaned up.

But there are also some things that need to be updated due to changes in ET.  Let's discuss tomorrow.

emmarogge committed:
Address reviewer feedback.

wnojopra commented on nextflow/README.md:
```diff
@@ -0,0 +1,6 @@
+
+This example shows how to get started with [Nextflow](https://www.nextflow.io) on Enterprise Terra.
+
+- The [nextflow_examples.ipynb](./nextflow_examples.ipynb) notebook shows how to configure Nextflow for in your Terra cloud environment, how to install and run pipelines from [nf-core](https://nf-co.re/rnaseq) workflows, and how to leverage the [Google Life Sciences API](https://cloud.google.com/life-sciences/docs/concepts/introduction) as the workflow process executor.
```
Minor nit - I'd break this line up into a bulleted list

emmarogge replied:
done

wnojopra commented on nextflow/nextflow_examples.ipynb:
```diff
@@ -0,0 +1,1097 @@
+    "    if not (os.getenv('GOOGLE_CLOUD_PROJECT')\n",
+    "            and os.getenv('GOOGLE_SERVICE_ACCOUNT_EMAIL')\n",
+    "            and os.getenv('TERRA_USER_EMAIL')) :\n",
+    "        raise Exception('Restart your TVC Cloud Environment so that the TVC environment variables become available.')       \n",
```
Nit - I like when these "check" functions output something like "Environment variables X Y Z set properly" in the positive case. Makes it real clear what the code actually did something when run.

emmarogge replied:
updated!

emmarogge committed:
Changes to address reviewer feedback

Changes include:
- Remove mention of SSH key from workspace description step 0.
- Move nf-core tool install to Setup section.
- Refactor nf-core tool install section to prompt user to restart kernel
before importing package.
- Remove links to parent workspace previews and include content about
how to accesss previews in user's workspace Resources tab.
- Update README content to point user to Resources panel and use
bulleted list.

wnojopra commented on nextflow/nextflow_examples.ipynb:
```diff
@@ -0,0 +1,1118 @@
+    "    import nf_core\n",
+    "    print(\"nf-core is already installed\")\n",
+    "except:\n",
+    "    print(\"Please restart the kernel before importing...\")"
```
Lets move these import cells to the very top. Right now, if I run the cells in order and then restart my kernel, I'll lose my BUCKET_REFERENCE variable.

emmarogge replied:
fixed!

wnojopra commented on nextflow/workspace_description.md:
```diff
@@ -0,0 +1,48 @@
+
+Launch the environment once it's running, and then run the notebook [workspace_setup.ipynb](https://github.com/DataBiosphere/terra-axon-examples/blob/main/workspace_setup.ipynb). Its repo, `terra-axon-examples`, which is defined as a workspace Git repository, should be automatically cloned to your Enterprise Terra Cloud Environment, and you should be able to navigate to the notebook in the Jupyte ...
+
+> **If you do not see the `terra-axon-examples` subdirectory**, open a Terminal window on the notebook server (under the **File** menu) and run:
```
Something changed slightly here in the last commit, because this now looks like 
![image](https://github.com/DataBiosphere/terra-axon-examples/assets/6879774/df9b8a46-379f-467e-a401-e6cdfb07bfa9)

emmarogge replied:
That's strange. This is how the content renders in the Jupyter Markdown preview of the file: https://screenshot.googleplex.com/9Ycw9AsLNkVMxD4. I also checked with an online Markdown editor just to validate and saw this: https://screenshot.googleplex.com/55D6robcEyW7Xtf.

I'll look into whether this is a nuance of GitHub Markdown.

emmarogge replied:
Fixed.

wnojopra commented on nextflow/workspace_description.md:
```diff
@@ -0,0 +1,48 @@
+## Step 4: View MultiQC reports produced.
+
+Running either or both examples will result in the creation of a MultiQC report. To preview what reports can be expected to look like, navigate to the workspace resource folder "Report previews" and click to view:
+1. [Example 1 MultiQC report preview](https://terra-preprod-ui-terra.api.verily.com/workspaces/getting-started-with-workflows-workspace/resources/6cce61b8-4dff-4342-934b-243134d91cac/example_results/multiqc_report.html)
```
I think you'll want to do the same thing here with these links that you did with the notebook links

emmarogge replied:
Fixed!

wnojopra reviewed (changes requested):
Nice! Almost there

emmarogge committed:
Updates to address reviewer comments

Changes:
- Remove preview links from 4th section of workspace description.
- Fix formatting of instruction for manually cloning repo.
- Refactor Notebook Setup section to put nf-core install first and make
content more straightforward.

wnojopra reviewed (approved):
LGTM

emmarogge committed:
Update filepaths to referenced Git repos to include the new /repos directory.

emmarogge committed:
Provide feedback when environment variables available as expected
